src/topology/test/backpressure.rs

```diff
@@ -0,0 +1,273 @@
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+
+use tokio::time::Duration;
+
+use crate::config::SinkOuter;
+
+use crate::topology::builder::PIPELINE_BUFFER_SIZE;
+use crate::{config::Config, test_util::start_topology};
+
+use vector_core::buffers::{BufferConfig, BufferType, WhenFull};
+
+pub const MEMORY_BUFFER_DEFAULT_MAX_EVENTS: usize = 500;
+
+#[tokio::test]
+async fn serial_backpressure() {
+    let mut config = Config::builder();
+
+    let events_to_sink = 100;
+
+    let expected_sourced_events =
+        events_to_sink + MEMORY_BUFFER_DEFAULT_MAX_EVENTS + PIPELINE_BUFFER_SIZE + 3;
+
+    let source_counter = Arc::new(AtomicUsize::new(0));
+    config.add_source(
+        "in",
+        test_source::TestBackpressureSourceConfig {
+            counter: source_counter.clone(),
+        },
+    );
+    config.add_sink(
+        "out",
+        &["in"],
+        test_sink::TestBackpressureSinkConfig {
+            num_to_consume: events_to_sink,
+        },
+    );
+
+    let (_topology, _crash) = start_topology(config.build().unwrap(), false).await;
+
+    // allow the topology to run
+    tokio::time::sleep(Duration::from_millis(500)).await;
+
+    let sourced_events = source_counter.load(Ordering::Relaxed);
+
+    assert_eq!(sourced_events, expected_sourced_events);
+}
+
+#[tokio::test]
+async fn default_fan_out() {
```

Comment:
I think all of these tests could probably benefit from some comments about the behavior they are asserting since it takes a close reading to ascertain.